bundles/statistics/statsgrid2016/components/IndicatorParametersHandler.js

```diff
@@ -0,0 +1,152 @@
+    this.indicators = null;
+
+    Oskari.makeObservable(this);
+}, {
+    getData: function ( datasrc, indicators, elements ) {
+
+        this.datasource = datasrc;
+        this.indicators = indicators;
+
+        if ( Array.isArray( indicators ) ) {
+            this.handleMultipleIndicators(datasrc, indicators, elements);
+            return;
+        }
+        this.handleSingleIndicator(datasrc, indicators, elements);
+    },
+    handleSingleIndicator: function(datasrc, indId, elements) {
+        var me = this;
+        var errorService = me.service.getErrorService();
+        var locale = this.locale;
+        var panelLoc = locale.panels.newSearch;
+
+        this.service.getIndicatorMetadata(datasrc, indId, function (err, indicator) {
+            if (elements.dataLabelWithTooltips) {
+                elements.dataLabelWithTooltips.find('.tooltip').hide();
+            }
+            if (err) {
+                // notify error!!
+                errorService.show(locale.errors.title, locale.errors.indicatorMetadataError);
+                return;
+            }
+
+            // selections
+            var selections = [];
+            var combinedValues = {};
+
+            indicator.selectors.forEach(function (selector, index) {
+                selector.allowedValues.forEach(function (val) {
+                    if ( !combinedValues[selector.id] ) {
+                        combinedValues[selector.id] = [];
+                    }
+
+                    var name = val.name || val.id || val;
+                    val.title = val.name;
+                    var optName = (panelLoc.selectionValues[selector.id] && panelLoc.selectionValues[selector.id][name]) ? panelLoc.selectionValues[selector.id][name] : name;
+
+                    var valObject = {
+                        id: val.id || val,
+                        title: optName
+                    };
+                    combinedValues[selector.id].push(valObject);
+                    selections.push(valObject);
+                });
+            });
+
+            if (indicator.regionsets.length === 0) {
+                errorService.show(locale.errors.title, locale.errors.regionsetsIsEmpty);
+            }
+
+            // Add margin if there is selections
+            if (indicator.selectors.length > 0) {
+                //regionSelect.container.addClass('margintop');
+            } else {
+                errorService.show(locale.errors.title, locale.errors.indicatorMetadataIsEmpty);
+            }
+
+            var data = {
+                datasrc: me.datasource,
+                indicators: me.indicators,
+                selectors: selections,
+                regionSet: indicator.regionsets,
+                values: combinedValues
+            }
+
+            me.trigger('Data.Loaded', data);
+        });
+    },
+    handleMultipleIndicators: function (datasrc, indicators, elements) {
+        indicators = indicators.filter( function (n) { return n != "" } );
```

Comment:
In my opinion this could be implemented in a more maintainable and simpler code if handleSingleIndicator() took a callback as parameter, handleMultipleIndicators() would call it in a loop and do the "extra stuff" after all the data has been fetched based on the values object for the single indicators.

Reply:
Please refactor towards this ^